### PR TITLE
[3.10] Handle the removal of the FTP support from Joomla 4 for com_joomlaupdate

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

Handle the removal of the FTP support from Joomla 4 for com_joomlaupdate. At the time of the upload we can just add a message given that we dont know the version you are updating to at this point.

### Testing Instructions

- Install 3.10
- point to an Joomla 4 package
- extra: please make sure that on 3.x the option will be showed correctly

### Actual result BEFORE applying this Pull Request

You still have the FTP options

### Expected result AFTER applying this Pull Request

The FTP options are gone

### Documentation Changes Required

None

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_joomlaupdate.ini [hunk 1]
```diff
@@ -90,6 +90,8 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_DB_NOT_SUPPORTED="Your database type is not suppor
 COM_JOOMLAUPDATE_VIEW_DEFAULT_DB_NOT_SUPPORTED_DESC="An update to Joomla %1$s was found, but your current database type is not supported by the new version.<br>For further details take a look at <a href="_QQ_"https://downloads.joomla.org/technical-requirements"_QQ_">the minimum requirements for Joomla %1$s</a>."
 COM_JOOMLAUPDATE_VIEW_DEFAULT_PHP_VERSION_NOT_SUPPORTED="Your PHP version is not supported"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_PHP_VERSION_NOT_SUPPORTED_DESC="An update to Joomla %1$s was found, but your currently installed PHP version does not match <a href="_QQ_"https://downloads.joomla.org/technical-requirements"_QQ_">the minimum requirements for Joomla %1$s</a>."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_NOTICE="Warning"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_NOTICE_MESSAGE="The FTP method will not work when you are upgrading to Joomla 4.0.0 or later."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: better to say "is not supported" than "will not work"

Comment (zero-24) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]:
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_NOTICE_MESSAGE="The FTP method will is not supported when you are upgrading to Joomla 4.0.0 or later."
```

Comment (zero-24) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: Something like that?

Comment (brianteeman) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: "is not supported"

Comment (zero-24) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: Thanks updated the suggestion.

Comment (ghost) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: The FTP method ~will~ is not supported ...

File: administrator/language/en-GB/en-GB.com_joomlaupdate.ini [hunk 2]
```diff
@@ -90,6 +90,8 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_DB_NOT_SUPPORTED="Your database type is not suppor
 COM_JOOMLAUPDATE_VIEW_DEFAULT_DB_NOT_SUPPORTED_DESC="An update to Joomla %1$s was found, but your current database type is not supported by the new version.<br>For further details take a look at <a href="_QQ_"https://downloads.joomla.org/technical-requirements"_QQ_">the minimum requirements for Joomla %1$s</a>."
 COM_JOOMLAUPDATE_VIEW_DEFAULT_PHP_VERSION_NOT_SUPPORTED="Your PHP version is not supported"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_PHP_VERSION_NOT_SUPPORTED_DESC="An update to Joomla %1$s was found, but your currently installed PHP version does not match <a href="_QQ_"https://downloads.joomla.org/technical-requirements"_QQ_">the minimum requirements for Joomla %1$s</a>."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_NOTICE="Warning"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_NOTICE_MESSAGE="The FTP method will is not supported when you are upgrading to Joomla 4.0.0 or later."
```
Comment (zero-24) [inline administrator/language/en-GB/en-GB.com_joomlaupdate.ini]:
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_NOTICE_MESSAGE="The FTP method is not supported when you are upgrading to Joomla 4.0.0 or later."
```

File: administrator/language/en-GB/en-GB.com_joomlaupdate.ini (hunk 1 above)
Comment (zero-24) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: Fixed by: https://github.com/joomla/joomla-cms/pull/34253/commits/1dd5d0a87af9e3a8b24e5e2eb45fb4b4bedd18b4

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 1dd5d0a87af9e3a8b24e5e2eb45fb4b4bedd18b4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34253">issues.joomla.org/tracker/joomla-cms/34253</a>.</sub>

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 1dd5d0a87af9e3a8b24e5e2eb45fb4b4bedd18b4

After patch applied the red marked `Installation method` isn't available:

image


<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34253">issues.joomla.org/tracker/joomla-cms/34253</a>.</sub>

**Comment by richard67** [comment]
As this PR here has 2 successful tests I wanted to set it RTC.

But there is something I don't understand.

When reading the code, the version compare here results in true if the first argument is greater or equal (because operator is '>=") than the second, and otherwise false, see PHP docs here: https://www.php.net/manual/en/function.version-compare.php

But I have just tested that the PR works like it should work and as it is described, and so it work vice versa than I would expect when reading the code.

I see 2 possible explanations for that:
1. I understand something completely wrong or am blind, or
2. The version compare at that place and also the search for an update has problems with our extra versions like "beta8-dev" which are part of the current version when testing on a current 3.10-dev branch or a 3.10 nightly build and also part of the update version when using the custom URL of the nightly builds.

Can someone else check and help me out?

**Comment by brianteeman** [comment]
I am exactly the same as you @richard67 that's why I installed it to test because I was sure it was wrong ;)

**Comment by richard67** [comment]
I'd still like to understand why it works or where I am wrong.

**Comment by richard67** [comment]
I have tested this item :red_circle: unsuccessfully on 1dd5d0a87af9e3a8b24e5e2eb45fb4b4bedd18b4<br /><br />When using current 3.10-dev branch or a current 3.10 nightly, which both have an extra version, and you use the custom update URLs for nightlies of 3.10 or 4.0 for testing, then this PR behaves as described, even if the version compare in the code is wrong.

But if you apply this patch to a 3.10 Alpha 4 and then attempt to update to Alpha 5, the method selection is not shown, but it should be.

When changing the version compare to "version_compare($this->updateInfo['latest'], '4.0.0', '<')", it still doesn't work as expected because the version compare seems to be confused by the extra versions.

When changing the version compare to "version_compare($this->updateInfo['latest'], '4', '<')", it works as expected. This is the same issue which we once had in the "Required PHP & Database Settings" to show the check for database supported only when updating to 4.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34253">issues.joomla.org/tracker/joomla-cms/34253</a>.</sub>

File: administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php [hunk 3]
```diff
@@ -64,54 +64,57 @@
...
-		<tr id="row_ftp_directory" <?php echo $this->ftpFieldsDisplay; ?>>
-			<td>
-				<?php echo JText::_('COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_DIRECTORY'); ?>
-			</td>
-			<td>
-				<input type="text" name="ftp_root" value="<?php echo $this->ftp['directory']; ?>" />
-			</td>
-		</tr>
+		<?php // Hide FTP settings when updating to Joomla 4 given that the supporting code has been dropped there ?>
+		<?php if (version_compare($this->updateInfo['latest'], '4.0.0', '>=')) : ?>
```
Comment (richard67) [inline administrator/components/com_joomlaupdate/views/default/tmpl/default_update.php]:
```suggestion
		<?php if (version_compare($this->updateInfo['latest'], '4', '<')) : ?>
```
See the comment in my test result.

**Comment by zero-24** [comment]
Merged thanks

**Comment by richard67** [comment]
I have not tested this item.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34253">issues.joomla.org/tracker/joomla-cms/34253</a>.</sub>

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 9cfc343f28b720820b8e7434857d5d3cd375a7ea<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34253">issues.joomla.org/tracker/joomla-cms/34253</a>.</sub>

**Comment by richard67** [comment]
@sandramay0905 @brianteeman Could you test again if it still works for you? Thanks in advance.

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 9cfc343f28b720820b8e7434857d5d3cd375a7ea<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34253">issues.joomla.org/tracker/joomla-cms/34253</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34253">issues.joomla.org/tracker/joomla-cms/34253</a>.</sub>

**Comment by infograf768** [comment]
tks

**Pull request merged**